### PR TITLE
Added Rings implementation documentation

### PR DESCRIPTION
**Pull request by yradsmikham**

### CODE REVIEW

File: gitops/azure-devops/ImageTagRelease.md [hunk 1]
```diff
@@ -208,3 +208,36 @@ Resource-Manifest-Repo
 Further reference:
 + [GitOps Pipeline Thinking](PipelineThinking.md)
 + [Manifest Generation Pipeline](README.md)
+
+
+## Create a service connection to ACR
+
+In order to add the ACR as an artifact or reference it in azure-pipelines.yml for a Docker build task, it's a good idea to add a service connection.
```
Comment (evanlouie) [inline gitops/azure-devops/ImageTagRelease.md]: > it's a good idea to add a service connection

Good idea? or required?

File: gitops/rings/README.md [hunk 2]
```diff
@@ -0,0 +1,144 @@
+# Rings
```
Comment (evanlouie) [inline gitops/rings/README.md]: Do we have any high level docs explaining the concepts of rings we inject in the intro?

File: gitops/rings/README.md [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+**“Flatpack” HLD Repository**: A git repository that maintains a High Level Definition for all Services and Revisions that are intended to be run on the Cluster.
+
+**“Materialized” Manifest Repository**: this git repository acts as our canonical source of truth for Flux – the in-cluster component that pulls and applies Kubernetes manifests rendered from the Flatpack HLD repository.
+
+### Ring.yaml
+As a ring is considered to be strictly a revision of a microservice, we need a way to configure the ingress controller to route to the microservice revision a user belongs to. We achieve this by providing a `ring.yaml` file in our helm chart, which is an abstraction on Kubernetes and Traefik primitives.
+
+An example of ring.yaml:
+
+```
```
Comment (evanlouie) [inline gitops/rings/README.md]: change to "```yaml" to get yaml highlighting

File: gitops/rings/README.md [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+
+### 2. Image Tag Release Pipeline
+
+The [Image Tag Release Pipeline](https://github.com/microsoft/bedrock/blob/rings/gitops/azure-devops/ImageTagRelease.md), which is a core component of the Bedrock CI/CD workflow, will acknowledge the creation of a new Ring when a git branch is created. Like any other commit, it will trigger the build for the Image Tag Release Pipeline. Recall that the pipeline will execute a build to build and push a Docker image using a new image tag. Then, it will initiate the Release pipeline, where a Pull Request will be created to have Fabrikate  update the image tag in the Fabrikate definitions for the Service HLD repo, __in addition to__ to adding a new subcomponent (via `fab add` command) to the `component.yaml` (shown below) for the new git branch, if it does not already exists. In the `component.yaml`, we use a path based selector to identify where in the source code repository the Helm chart exists.
+
+```
```
Comment (evanlouie) [inline gitops/rings/README.md]: change to "```yaml" to get yaml highlighting

File: gitops/rings/README.md [hunk 5]
```diff
@@ -0,0 +1,144 @@
...
+
+A developer on a project must manually engage a Pull Request merge in order to gate access for a Ring to production.
+
+### 4. Manifest Generation Pipeline
+
+When the pull request merged by another developer into the `master` branch of the Service HLD repository, the [Manifest Generation Pipeline](https://github.com/microsoft/bedrock/blob/master/gitops/azure-devops/ManifestGeneration.md) initiates. The difference in this Manifest Generation pipeline is that it will execute the build using the Flatpack HLD repo as opposed to the Service HLD repo. The Flatpack HLD repo sources all known services that are intended to be run on a cluster via their representative HLD repositories.
+
+An example of the `component.yaml` in the Flatpack HLD repo:
+
+```
```
Comment (evanlouie) [inline gitops/rings/README.md]: change to "```yaml" to get yaml highlighting

File: gitops/rings/RingsImplementation.md [hunk 6]
```diff
@@ -0,0 +1,291 @@
...
+## Pre-requisites
+
+Make sure you have the Azure Container Registry setup before proceeding. You will also need to add a [service connection](../azure-devops/ImageTagRelease.md#Create-a-service-connection-to-ACR) that this pipeline can use to access the ACR.
+
+## SRC to ACR
+
+This is the first pipeline in the process which builds the source repository, runs any required tests, and pushes the built image to the container registry. You may refer to our [sample hello world rings](https://github.com/bnookala/hello-rings) repository located here.
+
+Create a new `azure-pipelines.yml` file in the source repository and add the following code to it:
+```
```
Comment (evanlouie) [inline gitops/rings/RingsImplementation.md]: change to "```yaml" to get yaml highlighting

File: gitops/rings/RingsImplementation.md [hunk 7]
```diff
@@ -0,0 +1,291 @@
...
+4. Add two build steps to this pipeline, one downloads the necessary pre-requisites and the second runs the release pipeline.
+
+    - **GitHub**
+
+        If you're using GitHub repositories, follow the steps below.
+
+        For the first build step, copy the following lines of code:
+
+
+        ```
```
Comment (evanlouie) [inline gitops/rings/RingsImplementation.md]: change to "```sh" to get bash highlighting

File: gitops/rings/RingsImplementation.md [hunk 8]
```diff
@@ -0,0 +1,291 @@
...
+        sudo add-apt-repository ppa:cpick/hub
+        sudo apt-get update
+        sudo apt-get install hub
+        ```
+
+        ![](./images/download_prereqs.png)
+
+        For the second step, copy the following lines of code:
+
+        ```
```
Comment (evanlouie) [inline gitops/rings/RingsImplementation.md]: change to "```sh" to get bash highlighting

File: gitops/rings/RingsImplementation.md [hunk 9]
```diff
@@ -0,0 +1,291 @@
...
+        hub pull-request -m "Updating service $branch_name"
+        ```
+
+        ![](./images/run_release.png)
+
+    - **Azure DevOps**
+
+        If you're using Azure DevOps repos, follow the steps below. For the first build step, copy the following lines of code:
+
+        ```
```
Comment (evanlouie) [inline gitops/rings/RingsImplementation.md]: change to "```sh" to get bash highlighting

File: gitops/rings/RingsImplementation.md [hunk 10]
```diff
@@ -0,0 +1,291 @@
...
+
+The Service HLD to Materialized Manifest repo pipeline is initiated by a pull request that the Image Tag Release pipeline will create. A user will need to merge the pull request in order to cause this pipeline to build.
+
+### FlatPack HLD to Materialized Manifest
+
+As described in the [Rings Model](./README.md) documentation, the idea of using a Flatpack HLD is to have a repository that maintains the High Level Definition for **all** services and revisions that are intended to be run on the cluster.
+
+It is important to note that the Flatpack HLD to manifest pipeline build does *not* run every time the CI/CD process is invoked. The FlatPack HLD is a unique HLD that is **only** deployed when a new Ring is to be added to the cluster. The following example is a Flatpack HLD that consists of two services, the `hello-rings` and the `ring-operator`, each service with its own ring.
+
+```
```
Comment (evanlouie) [inline gitops/rings/RingsImplementation.md]: change to "```yaml" to get yaml highlighting

File: gitops/rings/README.md [hunk 11]
```diff
@@ -0,0 +1,144 @@
...
+ * config
+   * common.yaml
+ * src
+   * Dockerfile
+   * ..
+ * azure-pipelines.yaml
+
+The `selector` field of your `ring.yaml` should match the `spec.selector.matchLabels` of the `deployment.yaml` file. This is because the Ring Operator will create the service that matches the labels on the `deployment.yaml`, and once the Ring is created, it knows to map to a specific version of a service and ingress routes based on these `selector` labels, and will control access control to the service.
+
+### Ring Operator
```
Comment (evanlouie) [inline gitops/rings/README.md]: We should probably add a stub or something noting that this is currently BYO (unless we plan on open sourcing the ring operator for traefik as well)

File: gitops/rings/RingsImplementation.md [hunk 12]
```diff
@@ -0,0 +1,291 @@
...
+
+        # Download az CLI
+        sudo apt-get update
+        sudo apt-get install azure-cli
+        az extension add --name azure-devops
+        ```
+
+        For the second build step:
+
+        ```
```
Comment (evanlouie) [inline gitops/rings/RingsImplementation.md]: change to "```sh" to get bash highlighting

File: gitops/azure-devops/ImageTagRelease.md (hunk 1 above)
Comment (yradsmikham) [reply gitops/azure-devops/ImageTagRelease.md]: Required.

File: gitops/rings/README.md (hunk 11 above)
Comment (yradsmikham) [reply gitops/rings/README.md]: As in BYO ring operator... or?

File: gitops/rings/README.md [hunk 13]
```diff
@@ -0,0 +1,148 @@
...
+**Cluster HLD Repository**: A git repository that maintains a High Level Definition for all Services and Revisions that are intended to be run on the Cluster.
+
+**“Materialized” Manifest Repository**: this git repository acts as our canonical source of truth for Flux – the in-cluster component that pulls and applies Kubernetes manifests rendered from the Cluster HLD repository.
+
+### Ring.yaml
+As a ring is considered to be strictly a revision of a microservice, we need a way to configure the ingress controller to route to the microservice revision a user belongs to. We achieve this by providing a `ring.yaml` file in our helm chart, which is an abstraction on Kubernetes and Traefik primitives.
+
+An example of ring.yaml:
+
+```yaml
```
Comment (bnookala) [inline gitops/rings/README.md]: This file is actually the ring.yaml that gets output via the helm chart - not the file that gets consumed.

The file that we want for configuration is actually this:

```
serviceName: "languages"
majorVersion: "v1"
ringName: "48883-add-search"
group: "CMS-CORE"
contact: cms-core-dl@contoso.com
branch: feature/48883-add-search
initialUsers:
  - bill@contoso.com
  - ted@contoso.com
deploy: true

application:
  replicaCount: 10
  ports:
    - 8080
```

File: gitops/rings/README.md [hunk 14]
```diff
@@ -0,0 +1,148 @@
...
+The `ring.yaml` pairs with a `deployment.yaml` within the helm chart of your `src` repository as follows:
+
+**Source Repository**:
+ * chart
+   * templates
+       * ring.yaml
+       * deployment.yaml
+   * Chart.yaml
+   * values.yaml
+ * config
```
Comment (bnookala) [inline gitops/rings/README.md]: we need to remove this directory actually as it's from the HLD that used to be checked into this repo.

File: gitops/rings/README.md [hunk 15]
```diff
@@ -0,0 +1,148 @@
...
+   * Dockerfile
+   * ..
+ * azure-pipelines.yaml
+
+The `selector` field of your `ring.yaml` should match the `spec.selector.matchLabels` of the `deployment.yaml` file. This is because a "Ring Operator" will create the service that matches the labels on the `deployment.yaml`, and once the Ring is created, it knows to map to a specific version of a service and ingress routes based on these `selector` labels, and will control access control to the service.
+
+### Ring Operator
+The ring.yaml is consumed by a custom resource controller, which we call the Ring Controller. The Ring Controller sets up two resources on the cluster that map traffic to the proper service revision: a Traefik Ingress Route that maps path and headers to a Kubernetes service, and a Kubernetes service that maps to the microservice deployment.
+
+**DISCLAIMER:** Currently, Bedrock does *not* provide an open-sourced "Ring Operator", or any tool that can manage ingress routes to a specific service.
```
Comment (bnookala) [inline gitops/rings/README.md]: We're working on getting this open sourced.

**Comment by bnookala** [comment]
@yradsmikham this is a really good start to documenting what we need - I'll work with you on getting some of the knowledge filled in here.

**Comment by andrebriggs** [comment]
@yradsmikham I've been thinking about how we can incorporate this into the [high level GitOps pipeline description](https://github.com/microsoft/bedrock/blob/master/gitops/PipelineThinking.md) we have to maintain some continuity in the documentation.

Perhaps we can talk about a ring pipeline augmentation in a more abstract manner if we aren't providing/maintaining a Rings Operator application. I'm thinking we leave it at diagrams. 

Additionally we should point out the impetus of why you might consider this pipeline augmentation up front. That way folks can immediately know if the situation applies to them. 

If I'm understanding correctly, a developer would consider this pipeline augmentation when the following are true:

- You want to practice Bedrock GitOps
- You want a specific application/microservice rollout behavior (e.g. similar to blue/green or canary)
- You don't want to use a service mesh
- You have an application that will serve to replicate service behaviors ( BYO Ring Operator application)
 
Thoughts? @evanlouie @bnookala @samiyaakhtar

**Comment by yradsmikham** [comment]
Closing this PR because it appears that the Ring Operator implementation using Traefik (v2.0) is currently blocked. We cannot open source a Ring Operator if there is not a working ingress operator that is compatible with it.

**Pull request closed without merging**